File: programmers/여행경로/Programmers43164.java
import java.util.*;

public class Programmers43164 {
    public String[] solution(String[][] tickets) {
        Map<String, PriorityQueue<String>> map = new HashMap<>();
        for (String[] ticket : tickets) {
            PriorityQueue<String> dest = map.getOrDefault(ticket[0], new PriorityQueue<>());
            dest.add(ticket[1]);
            map.put(ticket[0], dest);
        }
        return dfs(map, new LinkedList<>(), "ICN");
    }

    private String[] dfs(Map<String, PriorityQueue<String>> map, List<String> visit, String start) {
        while (map.containsKey(start) && !map.get(start).isEmpty()) {
            dfs(map, visit, map.get(start).poll());
        }
        visit.add(0, start);
        return visit.toArray(new String[0]);
    }
}

File: programmers/여행경로/Programmers43164.kt
package com.example.algorithm

import java.util.PriorityQueue

class Programmers43164 {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val map = mutableMapOf<String, PriorityQueue<String>>()
        for (ticket in tickets) {
            map.putIfAbsent(ticket[0], PriorityQueue<String>())
            map[ticket[0]]!!.add(ticket[1])
        }
        val history = mutableListOf<String>()
        dfs(map, history, "ICN")
        return history.toTypedArray()
    }

    private fun dfs(
        map: MutableMap<String, PriorityQueue<String>>,
        visit: MutableList<String>,
        departure: String,
    ) {
        while (map.containsKey(departure) && !map[departure]!!.isEmpty()) {
            val arrival = map[departure]!!.poll()
            dfs(map, visit, arrival)
        }
        visit.add(0, departure)
    }
}
